docs/core/v3.0.rst

```diff
@@ -0,0 +1,1763 @@
+from a Zarr hierarchy. For a store to be compatible with this
+specification, it must support a set of operations defined in the `Abstract store
+interface`_ subsection. The store interface can be implemented using a
+variety of underlying storage technologies, described in the
+subsection on `Store implementations`_.
+
+
+.. _abstract-store-interface:
+
+Abstract store interface
+------------------------
+
+The store interface is intended to be simple to implement using a
+variety of different underlying storage technologies. It is defined in
+a general way here, but it should be straightforward to translate into
+a software interface in any given programming language. The goal is
+that an implementation of this specification could be modular and
+allow for different store implementations to be used.
+
+The store interface defines a set of operations involving `keys` and
+`values`. In the context of this interface, a `key` is any string
+containing only characters in the ranges ``a-z``, ``A-Z``, ``0-9``, or
+in the set ``/.-_``, where the final character is **not** a ``/``
+character. A `value` is any sequence of bytes.
+
+It is assumed that the store holds (`key`, `value`) pairs, with only
+one such pair for any given `key`. I.e., a store is a mapping from
+keys to values. It is also assumed that keys are case sensitive, i.e.,
+the keys "foo" and "FOO" are different.
+
+To read and write partial values, a `range` specifies two integers
+`range_start` and `range_length`, that specify a part of the value
+starting at byte `range_start` (inclusive) and having a length of
+`range_length` bytes. `range_length` may be none, indicating all
+available data until the end of the referenced value. For example
+`range` ``[0, none]`` specifies the full value. Stores that do not
+support partial access can still answer the requests using cutouts
+of full values. It is recommended that the implementation of the
+``get_partial_values``, ``set_partial_values`` and
+``erase_values`` methods is made optional, providing fallbacks
+for them by default. However, it is recommended to supply those operations
+where possible for efficiency. Also, the ``get``, ``set`` and ``erase``
+can easily be mapped onto their `partial_values` counterparts.
+Therefore, it is also recommended to supply fallbacks for those if the
+`partial_values` operations can be implemented.
+An entity containing those fallbacks could be named ``StoreWithPartialAccess``.
+
+The store interface also defines some operations involving
+`prefixes`. In the context of this interface, a prefix is a string
+containing only characters that are valid for use in `keys` and ending
+with a trailing ``/`` character.
+
+The store operations are grouped into three sets of capabilities:
+**readable**, **writeable** and **listable**. It is not necessary for
+a store implementation to support all of these capabilities.
+
+A **readable store** supports the following operation:
+
+@@TODO add bundled & partial access
```

Comment:
Noticed when reading through this recently that we had some TODOs remaining. Would it make sense to convert these to issues and remove the TODOs? Or do we want to do something else for these?

Reply:
Wondering if it would be more streamlined if we try and deal with any remaining TODOs within this PR. I'd suggest using a code review and adding a line comment under any TODOs so we can discuss and resolve here.

Reply:
In this case the TODO seems to be done, I think I left this in accidentally. `get_partial_values` as described below adds the bundled & partial access already.